DigitalLearningSolutions.Data/Services/EmailService.cs

```diff
@@ -0,0 +1,127 @@
+﻿namespace DigitalLearningSolutions.Data.Services
+{
+    using System;
+    using DigitalLearningSolutions.Data.Factories;
+    using DigitalLearningSolutions.Data.Models.Email;
+    using MimeKit;
+
+    public interface IEmailService
+    {
+        void SendEmail(Email email);
+        // make class of email details with To, CC, (BCC?) as string lists, and body as BodyBuilder
+
+    }
+
+
+    public class EmailService : IEmailService
+    {
+        private readonly IConfigService configService;
+        private readonly ISmtpClientFactory smtpClientFactory;
+
+        public EmailService (
+            IConfigService configService,
+            ISmtpClientFactory smtpClientFactory
+        )
+        {
+            this.configService = configService;
+            this.smtpClientFactory = smtpClientFactory;
+        }
+
+        public void SendEmail(Email email)
+        {
+            var mailServerUsername = configService.GetConfigValue(ConfigService.MailUsername);
+            var mailServerPassword = configService.GetConfigValue(ConfigService.MailPassword);
+            var mailServerAddress = configService.GetConfigValue(ConfigService.MailServer);
+            var mailServerPortString = configService.GetConfigValue(ConfigService.MailPort);
+            var mailSenderAddress = configService.GetConfigValue(ConfigService.MailFromAddress);
+            if (
+                mailServerUsername == null
+                || mailServerPassword == null
+                || mailServerAddress == null
+                || mailServerPortString == null
+                || mailSenderAddress == null
+            )
+            {
+                var errorMessage = GenerateConfigValueMissingMessage(
+                    mailServerUsername,
+                    mailServerPassword,
+                    mailServerAddress,
+                    mailServerPortString
+                );
+                throw new ConfigValueMissingException(errorMessage);
+            }
+
+            var mailServerPort = int.Parse(mailServerPortString);
+
+            MimeMessage message = CreateMessage(email, mailSenderAddress);
+
+            try
+            {
+                using var client = smtpClientFactory.GetSmtpClient();
+                client.Timeout = 10000;
+                client.Connect(mailServerAddress, mailServerPort);
+
+                client.Authenticate(mailServerUsername, mailServerPassword);
+
+                client.Send(message);
+                client.Disconnect(true);
+            }
+            catch 
```

Comment:
Are we going to keep the "catch" here and swallow any errors?

Reply:
The empty catch was implemented in the EmailService at the same place it was in the SendSMTPMessage() in the NotificationService - it was brought over so that whatever originally used SendSMTPMessage() would experience the same functionality and would not break if an unexpected exception was thrown. I think it would be best to not swallow the errors, but it might require additional handling in the NotificationService (the PasswordResetService works as expected when there is an exception thrown there, showing the user the "A problem has occurred" error page). Would that additional handling be something to cover as part of this ticket?

Reply:
Yes, we did discuss this! I will create another ticket for this error handling! :)